File: src/simulation/meta/EventDirectorSystem.js
import { BALANCE } from "../../config/balance.js";
import { EVENT_TYPE } from "../../config/constants.js";
import { enqueueEvent } from "../../world/events/WorldEventQueue.js";
import { emitEvent, EVENT_TYPES as BUS_EVENT_TYPES } from "./GameEventBus.js";

/**
 * v0.8.2 Round-6 Wave-2 (01d-mechanics-content Step 3) — EventDirectorSystem.
 *
 * Why this exists: WorldEventSystem already consumes events from
 * `state.events.queue` (BANDIT_RAID / TRADE_CARAVAN / ANIMAL_MIGRATION + the
 * three Wave-2 additions DISEASE_OUTBREAK / WILDFIRE / MORALE_BREAK), but the
 * only authoritative producer in the codebase is `EnvironmentDirectiveApplier`,
 * which depends on a live LLM directive. With proxy offline rate ≈100% in the
 * default benchmark/dev runs the queue stays empty and no proactive pressure
 * ever materialises. EventDirectorSystem fixes the gap with a deterministic,
 * weight-driven pump that pushes one event into the queue every
 * `BALANCE.eventDirectorBaseIntervalSec` (default 240s game-time).
 *
 * Determinism: uses `services.rng.next()` so 4-seed benchmark runs stay
 * reproducible. If services.rng is missing (legacy bootstrap path), falls
 * back to Math.random.
 *
 * Bandit-raid downgrade: if the rolled type is BANDIT_RAID but
 * `state.gameplay.raidEscalation` indicates the raid cooldown has not yet
 * elapsed (currentTick - lastRaidTick < intervalTicks), downgrade to the
 * next-highest-weight non-raid event so the cadence promise holds.
 */

const NON_RAID_FALLBACK_ORDER = Object.freeze([
  EVENT_TYPE.ANIMAL_MIGRATION,
  EVENT_TYPE.TRADE_CARAVAN,
  EVENT_TYPE.DISEASE_OUTBREAK,
  EVENT_TYPE.WILDFIRE,
  EVENT_TYPE.MORALE_BREAK,
]);

function ensureDirectorState(state) {
  state.gameplay ??= {};
  if (!state.gameplay.eventDirector || typeof state.gameplay.eventDirector !== "object") {
    state.gameplay.eventDirector = {
      lastDispatchSec: -Infinity,
      dayBudget: 0,
      history: [],
    };
  }
  return state.gameplay.eventDirector;
}

function rollEventType(rng) {
  const weights = BALANCE.eventDirectorWeights ?? {};
  const entries = [
    [EVENT_TYPE.BANDIT_RAID, Number(weights.banditRaid ?? 0)],
    [EVENT_TYPE.ANIMAL_MIGRATION, Number(weights.animalMigration ?? 0)],
    [EVENT_TYPE.TRADE_CARAVAN, Number(weights.tradeCaravan ?? 0)],
    [EVENT_TYPE.DISEASE_OUTBREAK, Number(weights.diseaseOutbreak ?? 0)],
    [EVENT_TYPE.WILDFIRE, Number(weights.wildfire ?? 0)],
    [EVENT_TYPE.MORALE_BREAK, Number(weights.moraleBreak ?? 0)],
  ];
  const total = entries.reduce((sum, [, w]) => sum + Math.max(0, w), 0);
  if (total <= 0) return EVENT_TYPE.ANIMAL_MIGRATION;
  let pick = rng() * total;
  for (const [type, weight] of entries) {
    pick -= Math.max(0, weight);
    if (pick <= 0) return type;
  }
  return entries[entries.length - 1][0];
}

function isRaidOnCooldown(state) {
  const esc = state.gameplay?.raidEscalation ?? null;
  const intervalTicks = Number(esc?.intervalTicks ?? BALANCE.raidIntervalBaseTicks ?? 3600);
  const currentTick = Number(state.metrics?.tick ?? 0);
  const lastRaidTick = Number(state.gameplay?.lastRaidTick ?? -9999);
  return (currentTick - lastRaidTick) < intervalTicks;
}

function downgradeRaid(rng) {
  // Re-roll over non-raid weights so a downgrade does not silently bias the
  // distribution toward the cheapest fallback.
  const weights = BALANCE.eventDirectorWeights ?? {};
  const map = {
    [EVENT_TYPE.ANIMAL_MIGRATION]: Number(weights.animalMigration ?? 0),
    [EVENT_TYPE.TRADE_CARAVAN]: Number(weights.tradeCaravan ?? 0),
    [EVENT_TYPE.DISEASE_OUTBREAK]: Number(weights.diseaseOutbreak ?? 0),
    [EVENT_TYPE.WILDFIRE]: Number(weights.wildfire ?? 0),
    [EVENT_TYPE.MORALE_BREAK]: Number(weights.moraleBreak ?? 0),
  };
  const total = Object.values(map).reduce((sum, w) => sum + Math.max(0, w), 0);
  if (total <= 0) return EVENT_TYPE.ANIMAL_MIGRATION;
  let pick = rng() * total;
  for (const type of NON_RAID_FALLBACK_ORDER) {
    pick -= Math.max(0, map[type] ?? 0);
    if (pick <= 0) return type;
  }
  return EVENT_TYPE.ANIMAL_MIGRATION;
}

function getTuning(eventType) {
  const tuning = BALANCE.eventDirectorTuning ?? {};
  // Map EVENT_TYPE string back to balance key.
  const keyMap = {
    [EVENT_TYPE.BANDIT_RAID]: "banditRaid",
    [EVENT_TYPE.ANIMAL_MIGRATION]: "animalMigration",
    [EVENT_TYPE.TRADE_CARAVAN]: "tradeCaravan",
    [EVENT_TYPE.DISEASE_OUTBREAK]: "diseaseOutbreak",
    [EVENT_TYPE.WILDFIRE]: "wildfire",
    [EVENT_TYPE.MORALE_BREAK]: "moraleBreak",
  };
  const key = keyMap[eventType] ?? "animalMigration";
  const entry = tuning[key] ?? {};
  return {
    durationSec: Number(entry.durationSec ?? 24),
    intensity: Number(entry.intensity ?? 1),
  };
}

export class EventDirectorSystem {
  constructor() {
    this.name = "EventDirectorSystem";
  }

  update(dt, state, services) {
    if (!state || !state.events) return;
    const director = ensureDirectorState(state);
    const nowSec = Number(state.metrics?.timeSec ?? 0);
    const intervalSec = Number(BALANCE.eventDirectorBaseIntervalSec ?? 240);

    // v0.10.2 PJ-followup-cadence-dampener R7 — opening-crisis safety net.
    // While the colony has no farms AND we're in the bootstrap window (<180s),
    // throttle event cadence to 2.5× baseline so saboteur draft / wildlife
    // events don't pull workers off harvest during the food crash window.
    // Disengages naturally the moment Plan-PL-terrain-min-guarantee or the
    // autopilot's first farm completes. Defense-in-depth complement to the
    // PL terrain floor for adversarial seeds that slip through the guarantee.
    const bootstrapWindow = (state.buildings?.farms ?? 0) === 0 && Number(state.metrics?.timeSec ?? 0) < 180;
    const effectiveIntervalSec = bootstrapWindow ? intervalSec * 2.5 : intervalSec;

    const lastDispatch = Number.isFinite(director.lastDispatchSec) ? director.lastDispatchSec : -Infinity;
    // First-run grace: after init the lastDispatchSec is -Infinity, so the
    // very first tick would otherwise dispatch immediately. Anchor the first
    // dispatch and offset by half-interval so the first event lands at
    // ~intervalSec/2 instead of a full interval out — v0.10.2 PJ-pacing P0
    // pulls first event from t=intervalSec to t=intervalSec/2 (45s at the
    // new 90s base interval) so the early game has visible motion. The
    // dampener applies to this offset too: during bootstrap the first event
    // lands at ~effectiveIntervalSec/2 (≈112s) instead of ~45s.
    if (!Number.isFinite(lastDispatch)) {
      director.lastDispatchSec = nowSec - effectiveIntervalSec * 0.5;
      return;
    }
    if (nowSec - lastDispatch < effectiveIntervalSec) return;

    const rng = typeof services?.rng?.next === "function"
      ? () => services.rng.next()
      : Math.random;

    let chosen = rollEventType(rng);
    if (chosen === EVENT_TYPE.BANDIT_RAID && isRaidOnCooldown(state)) {
      chosen = downgradeRaid(rng);
    }
    const { durationSec, intensity } = getTuning(chosen);
    enqueueEvent(state, chosen, {}, durationSec, intensity);

    // v0.10.2 PJ-pacing P0 — surface the dispatch on the player-visible
    // event log so chronicle/HUD listeners can react. Reuses the generic
    // EVENT_STARTED bus type with a `kind:"event_started"` discriminator
    // (no new mechanic — pure log channel).
    emitEvent(state, BUS_EVENT_TYPES.EVENT_STARTED, {
      kind: "event_started",
      eventType: chosen,
      intensity,
      durationSec,
    });

    director.lastDispatchSec = nowSec;
    director.dayBudget = Number(director.dayBudget ?? 0) + 1;
    director.history ??= [];
    director.history.unshift({ sec: Number(nowSec.toFixed(1)), type: chosen });
    if (director.history.length > 32) director.history.length = 32;

    // Push a trace line when debug.eventTrace is enabled (parity with
    // WorldEventSystem trace format).
    if (state.debug?.eventTrace) {
      state.debug.eventTrace.unshift(
        `[t=${nowSec.toFixed(1)}s] director dispatched ${chosen}`,
      );
      state.debug.eventTrace = state.debug.eventTrace.slice(0, 36);
    }
  }
}

File: project_utopia/simulation/meta/event_director_system.py
"""Deterministic event scheduler (port of ``EventDirectorSystem.js``).

This is the **scheduler** that pushes events into the world-event queue
on a fixed cadence — *not* the LLM ``EnvironmentDirector`` channel and
*not* :class:`project_utopia.world.events.WorldEventSystem` (which drains
the queue). The two systems intentionally live in different modules.
"""

from __future__ import annotations

from typing import Any

from project_utopia.app.rng import SeededRng
from project_utopia.world.events import EventType, WorldEventState, enqueue_event

from .game_event_bus import EVENT_TYPES, emit_event

__all__ = ["EventDirectorSystem", "EVENT_DIRECTOR_BALANCE"]


# Tunable knobs (mirrors ``BALANCE.eventDirector*`` in JS). Frozen dict so
# tests can monkey-patch a per-test override via ``dict(... | overrides)``
# without mutating module state.
EVENT_DIRECTOR_BALANCE: dict[str, Any] = {
    "base_interval_sec": 240.0,
    "weights": {
        EventType.BANDIT_RAID: 0.20,
        EventType.ANIMAL_MIGRATION: 0.25,
        EventType.TRADE_CARAVAN: 0.25,
        EventType.DISEASE_OUTBREAK: 0.10,
        EventType.WILDFIRE: 0.10,
    },
    "tuning": {
        EventType.BANDIT_RAID: {"duration_sec": 24.0, "intensity": 1.0},
        EventType.ANIMAL_MIGRATION: {"duration_sec": 24.0, "intensity": 1.0},
        EventType.TRADE_CARAVAN: {"duration_sec": 24.0, "intensity": 1.0},
        EventType.DISEASE_OUTBREAK: {"duration_sec": 24.0, "intensity": 1.0},
        EventType.WILDFIRE: {"duration_sec": 24.0, "intensity": 1.0},
    },
    "raid_interval_base_ticks": 3600,
    "history_cap": 32,
}

_NON_RAID_FALLBACK_ORDER: tuple[EventType, ...] = (
    EventType.ANIMAL_MIGRATION,
    EventType.TRADE_CARAVAN,
    EventType.DISEASE_OUTBREAK,
    EventType.WILDFIRE,
)


def _rng_next(rng: SeededRng | None) -> float:
    """RC3 B3-B7 nondeterminism fix: constant fallback, never ``random()``."""
    if rng is None:
        return 0.5
    return rng.next()


def _roll_event_type(rng: SeededRng | None, weights: dict[EventType, float]) -> EventType:
    """Weighted choice across ``weights`` using a single RNG draw."""
    entries = sorted(weights.items(), key=lambda kv: kv[0].value)
    total = sum(max(0.0, float(w)) for _, w in entries)
    if total <= 0.0:
        return EventType.ANIMAL_MIGRATION
    pick = _rng_next(rng) * total
    for evt_type, weight in entries:
        pick -= max(0.0, float(weight))
        if pick <= 0.0:
            return evt_type
    return entries[-1][0]


def _downgrade_raid(rng: SeededRng | None, weights: dict[EventType, float]) -> EventType:
    """Re-roll using non-raid weights when the bandit raid is on cooldown."""
    map_w = {t: weights.get(t, 0.0) for t in _NON_RAID_FALLBACK_ORDER}
    total = sum(max(0.0, float(w)) for w in map_w.values())
    if total <= 0.0:
        return EventType.ANIMAL_MIGRATION
    pick = _rng_next(rng) * total
    for evt_type in _NON_RAID_FALLBACK_ORDER:
        pick -= max(0.0, float(map_w.get(evt_type, 0.0)))
        if pick <= 0.0:
            return evt_type
    return EventType.ANIMAL_MIGRATION


def _ensure_director_state(state: dict[str, Any]) -> dict[str, Any]:
    gameplay = state.setdefault("gameplay", {})
    director = gameplay.get("event_director")
    if not isinstance(director, dict):
        director = {
            "last_dispatch_sec": float("-inf"),
            "day_budget": 0,
            "history": [],
        }
        gameplay["event_director"] = director
    return director


def _is_raid_on_cooldown(state: dict[str, Any], balance: dict[str, Any]) -> bool:
    gameplay = state.get("gameplay") or {}
    esc = gameplay.get("raid_escalation") or {}
    interval_ticks = float(
        esc.get("interval_ticks", balance.get("raid_interval_base_ticks", 3600))
    )
    metrics = state.get("metrics") or {}
    current_tick = float(metrics.get("tick", 0))
    last_raid_tick = float(gameplay.get("last_raid_tick", -9999))
    return (current_tick - last_raid_tick) < interval_ticks


class EventDirectorSystem:
    """Deterministic event scheduler.

    Pushes one event into ``state.events.queue`` (a
    :class:`WorldEventState`) every ``base_interval_sec`` of simulated
    time. Determinism is anchored on ``services.rng`` — if that isn't
    threaded through, we fall back to a constant ``0.5`` (RC3 B3 fix).
    """

    def __init__(self, balance: dict[str, Any] | None = None) -> None:
        self.name = "EventDirectorSystem"
        self.balance = dict(balance) if balance is not None else dict(EVENT_DIRECTOR_BALANCE)

    def update(
        self,
        dt: float,
        state: dict[str, Any],
        services: Any | None = None,
    ) -> None:
        del dt  # unused; cadence is keyed off state.metrics.timeSec
        events_obj = state.get("events")
        # Accept either dict (test fixtures) or WorldEventState dataclass.
        if events_obj is None:
            return
        director = _ensure_director_state(state)
        metrics = state.get("metrics") or {}
        now_sec = float(metrics.get("timeSec", 0.0))
        base_interval = float(self.balance.get("base_interval_sec", 240.0))
        bootstrap_window = (
            int((state.get("buildings") or {}).get("farms", 0)) == 0
            and now_sec < 180.0
        )
        effective_interval = base_interval * 2.5 if bootstrap_window else base_interval
        last_dispatch = float(director.get("last_dispatch_sec", float("-inf")))

        # First-tick anchor: align so the first dispatch lands at half-interval.
        if last_dispatch == float("-inf"):
            director["last_dispatch_sec"] = now_sec - effective_interval * 0.5
            return
        if (now_sec - last_dispatch) < effective_interval:
            return

        rng = getattr(services, "rng", None)
        weights = self.balance.get("weights", EVENT_DIRECTOR_BALANCE["weights"])
        chosen = _roll_event_type(rng, weights)
        if chosen == EventType.BANDIT_RAID and _is_raid_on_cooldown(state, self.balance):
            chosen = _downgrade_raid(rng, weights)
        tuning = self.balance.get("tuning", {}).get(chosen, {})
        duration_sec = float(tuning.get("duration_sec", 24.0))
        intensity = float(tuning.get("intensity", 1.0))

        # WorldEventState integration: prefer dataclass API when present,
        # else fall back to dict-of-lists for harness fixtures.
        if isinstance(events_obj, WorldEventState):
            enqueue_event(events_obj, chosen, {}, duration_sec, intensity)
        else:
            queue = events_obj.setdefault("queue", [])
            queue.append(
                {
                    "type": chosen.value if isinstance(chosen, EventType) else chosen,
                    "payload": {},
                    "duration_sec": duration_sec,
                    "intensity": intensity,
                }
            )

        emit_event(
            state,
            EVENT_TYPES["EVENT_STARTED"],
            {
                "kind": "event_started",
                "eventType": chosen.value if isinstance(chosen, EventType) else chosen,
                "intensity": intensity,
                "durationSec": duration_sec,
            },
        )

        director["last_dispatch_sec"] = now_sec
        director["day_budget"] = int(director.get("day_budget", 0)) + 1
        history = director.setdefault("history", [])
        history.insert(
            0,
            {
                "sec": round(now_sec, 1),
                "type": chosen.value if isinstance(chosen, EventType) else chosen,
            },
        )
        history_cap = int(self.balance.get("history_cap", 32))
        if len(history) > history_cap:
            del history[history_cap:]
